Implement ARIA id-ref reflection for ElementInternals

rniwa opened the pull request:
#### 6989a5b880ac9b18befc8e0c921bac4f778a2189
<pre>
Implement ARIA id-ref reflection for ElementInternals
<a href="https://bugs.webkit.org/show_bug.cgi?id=245328">https://bugs.webkit.org/show_bug.cgi?id=245328</a>

Reviewed by Manuel Rego Casasnovas.

This patch adds the support for id-ref ARIA attributes to ElementInternals.

* LayoutTests/accessibility/custom-elements/controls-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/controls.html: Added.
* LayoutTests/accessibility/custom-elements/describedby-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/describedby.html: Added.
* LayoutTests/accessibility/custom-elements/flowto-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/flowto.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-expected.txt:
* LayoutTests/accessibility/custom-elements/menuitem.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt:
* LayoutTests/platform/gtk/accessibility/custom-elements/describedby-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isModalElement const):
(WebCore::nodeHasRole):
(WebCore::AXObjectCache::updateRelationsIfNeeded): Add the support for defining relations via ElementInternals.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::getAttribute const):
(WebCore::AccessibilityObject::elementsFromAttribute const): Ditto.

* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp: Added.
(WebCore::getElementsArrayAttribute): Added.
(WebCore::JSElementInternals::ariaControlsElements const): Added.
(WebCore::JSElementInternals::ariaDescribedByElements const): Added.
(WebCore::JSElementInternals::ariaDetailsElements const): Added.
(WebCore::JSElementInternals::ariaErrorMessageElements const):
(WebCore::JSElementInternals::ariaFlowToElements const): Added.
(WebCore::JSElementInternals::ariaLabelledByElements const): Added.
(WebCore::JSElementInternals::ariaOwnsElements const): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible): Added.
(WebCore::CustomElementDefaultARIA::valueForAttribute const): Added the support for converting
WeakPtr&lt;Element&gt; and Vector&lt;WeakPtr&lt;Element&gt;&gt; to ID strings.
(WebCore::CustomElementDefaultARIA::elementForAttribute const): Added.
(WebCore::CustomElementDefaultARIA::setElementForAttribute): Added.
(WebCore::CustomElementDefaultARIA::elementsForAttribute const): Added.
(WebCore::CustomElementDefaultARIA::setElementsForAttribute): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.h:

* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::computeValueForAttribute): Extracted out of setAttributeWithoutSynchronization.
(WebCore::ElementInternals::setAttributeWithoutSynchronization):
(WebCore::ElementInternals::attributeWithoutSynchronization const):
(WebCore::ElementInternals::getElementAttribute const): Added.
(WebCore::ElementInternals::setElementAttribute): Added.
(WebCore::ElementInternals::getElementsArrayAttribute const): Added.
(WebCore::ElementInternals::setElementsArrayAttribute): Added.

* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:

Canonical link: <a href="https://commits.webkit.org/254709@main">https://commits.webkit.org/254709@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/985cac217a5a97a627899435da2c5de8c0be953a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89895 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99229 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/156261 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93902 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32933 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28352 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82228 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93550 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95541 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26153 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76696 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26075 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80858 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69078 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30680 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14941 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30418 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15882 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33879 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38841 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1411 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32595 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34966 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4033,6 +4034,13 @@ void AXObjectCache::updateRelationsIfNeeded()
         // Collect all possible origins, i.e., elements with non-empty relation attributes.
         for (const auto& attribute : relationAttributes()) {
             auto& idsString = element.attributeWithoutSynchronization(attribute);
+            if (idsString.isNull()) {
```
We need to remember to fix this code for shadow DOM later.

mrego replied:
Should we add a FIXME for later?

rniwa replied:
Sure, added.

mrego replied:
Where is the FIXME?

mrego commented on Source/WebCore/dom/CustomElementDefaultARIA.cpp:
```diff
@@ -40,17 +41,101 @@ void CustomElementDefaultARIA::setValueForAttribute(const QualifiedName& name, c
+}
+
+void CustomElementDefaultARIA::setElementForAttribute(const QualifiedName& name, Element* element)
+{
```
Nit: Add "ASSERT(element)"

rniwa replied:
That's not true. This value can be null.

mrego replied:
Right, sorry for the noise.

mrego commented on Source/WebCore/bindings/js/JSElementInternalsCustom.cpp:
```diff
@@ -0,0 +1,97 @@
+namespace WebCore {
+using namespace JSC;
+
+static JSValue getElementsArrayAttribute(JSGlobalObject& lexicalGlobalObject, const JSElementInternals& thisObject, const QualifiedName& attributeName)
```
It'd be really nice if we could share this code with `JSElement` somehow, to avoid duplicating it in both places.

rniwa replied:
Well, at some point, we need to start generating this code instead of writing as custom bindings code.

mrego replied:
Yeah, fair point.

mrego commented on Source/WebCore/dom/CustomElementDefaultARIA.cpp:
```diff
@@ -40,17 +41,101 @@ void CustomElementDefaultARIA::setValueForAttribute(const QualifiedName& name, c
+    }, [&](const Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>& elements) {
+        result.reserveCapacity(elements.size());
+        for (auto& weakElement : elements) {
+            if (RefPtr element = weakElement.get(); element && (!element->isConnected() || thisElement.isDescendantOrShadowDescendantOf(element->rootNode())))
```
It would be nice to reuse these checks somehow and avoid having to repeat them in the different methods.

rniwa replied:
Extracted it as a helper function: `isElementVisible`

mrego replied:
I guess the new changes haven't been pushed to the PR yet, or at least I cannot see them.

rniwa replied:
Oh weird. I pushed the new version.

mrego reviewed (approved):
LGTM, not sure if you also want to ask for review to someone with more experience on the accessibility code.

rniwa committed:
Implement ARIA id-ref reflection for ElementInternals
https://bugs.webkit.org/show_bug.cgi?id=245328

Reviewed by Manuel Rego Casasnovas.

This patch adds the support for id-ref ARIA attributes to ElementInternals.

* LayoutTests/accessibility/custom-elements/controls-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/controls.html: Added.
* LayoutTests/accessibility/custom-elements/describedby-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/describedby.html: Added.
* LayoutTests/accessibility/custom-elements/flowto-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/flowto.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-expected.txt:
* LayoutTests/accessibility/custom-elements/menuitem.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt:
* LayoutTests/platform/gtk/accessibility/custom-elements/describedby-expected.txt: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isModalElement const):
(WebCore::nodeHasRole):
(WebCore::AXObjectCache::updateRelationsIfNeeded): Add the support for defining relations via ElementInternals.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::getAttribute const):
(WebCore::AccessibilityObject::elementsFromAttribute const): Ditto.

* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp: Added.
(WebCore::getElementsArrayAttribute): Added.
(WebCore::JSElementInternals::ariaControlsElements const): Added.
(WebCore::JSElementInternals::ariaDescribedByElements const): Added.
(WebCore::JSElementInternals::ariaDetailsElements const): Added.
(WebCore::JSElementInternals::ariaErrorMessageElements const):
(WebCore::JSElementInternals::ariaFlowToElements const): Added.
(WebCore::JSElementInternals::ariaLabelledByElements const): Added.
(WebCore::JSElementInternals::ariaOwnsElements const): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible): Added.
(WebCore::CustomElementDefaultARIA::valueForAttribute const): Added the support for converting
WeakPtr<Element> and Vector<WeakPtr<Element>> to ID strings.
(WebCore::CustomElementDefaultARIA::elementForAttribute const): Added.
(WebCore::CustomElementDefaultARIA::setElementForAttribute): Added.
(WebCore::CustomElementDefaultARIA::elementsForAttribute const): Added.
(WebCore::CustomElementDefaultARIA::setElementsForAttribute): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.h:

* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::computeValueForAttribute): Extracted out of setAttributeWithoutSynchronization.
(WebCore::ElementInternals::setAttributeWithoutSynchronization):
(WebCore::ElementInternals::attributeWithoutSynchronization const):
(WebCore::ElementInternals::getElementAttribute const): Added.
(WebCore::ElementInternals::setElementAttribute): Added.
(WebCore::ElementInternals::getElementsArrayAttribute const): Added.
(WebCore::ElementInternals::setElementsArrayAttribute): Added.

* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:

Canonical link: https://commits.webkit.org/254709@main
